File: src/main/scala/com/github/alikemalocalan/tunnel/models/HttpRequest.scala
package com.github.alikemalocalan.tunnel.models

import java.net.URI

import io.netty.buffer.{ByteBuf, Unpooled}
import io.netty.handler.codec.http.HttpHeaders

case class HttpRequest(method: String,
                       uri: URI,
                       protocolVersion: String,
                       port: Int,
                       isHttps: Boolean,
                       headers: Option[HttpHeaders] = None,
                       payload: Option[String] = None) {

  def headersAsString: String = {
    val builder = new StringBuilder("")

    headers.foreach(_.entries().forEach { header =>
      builder.append(String.format("%s: %s\r\n", header.getKey, header.getValue.trim))
    })
    builder.toString
  }


  def getPath: String = {
    Option(uri.getPath).filter(_.nonEmpty) getOrElse ("/")
  }


  def host: String = uri.getHost

  override def toString: String = String.format("%s %s %s\r\n", method, getPath, protocolVersion) + headersAsString + "\r\n" + payload.getOrElse("")

  def toStringForHTTPS: String = String.format("CONNECT %s:%s %s\r\n", host, port, protocolVersion)

  def toByteBuf: ByteBuf = if (isHttps) Unpooled.EMPTY_BUFFER else Unpooled.wrappedBuffer(toString().getBytes())
}

File: src/main/kotlin/com/github/alikemalocalan/greentunnel4jvm/models/HttpRequest.kt
package com.github.alikemalocalan.greentunnel4jvm.models

import com.github.alikemalocalan.greentunnel4jvm.utils.DNSOverHttps
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import java.net.InetSocketAddress
import java.net.URI
import java.util.*

data class HttpRequest(
    val method: String,
    val uri: URI,
    val protocolVersion: String,
    val port: Int,
    val isHttps: Boolean,
    val headers: Optional<List<Pair<String, String>>> = Optional.empty(),
    val payload: Optional<String> = Optional.empty()
) {

    private fun headersAsString(): String =
        headers.map { headersNonNull ->
            "\r\n" + headersNonNull.reversed()
                .joinToString(separator = "\r\n", postfix = "\r\n") { header ->
                    String.format("%s: %s", header.first, header.second.trim())
                }
        }.orElseGet { "" }

    private fun path(): String =
        if (uri.path.isNullOrBlank()) "/"
        else uri.path


    fun host(): String = uri.host

    override fun toString(): String = String.format(
        "%s  %s  %s%s%s",
        method,
        path(),
        protocolVersion,
        headersAsString(),
        payload.map { payload -> "\r\n" + payload }.orElseGet { "" }
    )

    fun toByteBuf(): ByteBuf =
        if (isHttps) Unpooled.EMPTY_BUFFER
        else Unpooled.wrappedBuffer(toString().toByteArray(Charsets.UTF_8))

    fun toInetSocketAddress(): Optional<InetSocketAddress> =
        DNSOverHttps.lookUp(host())
            .map { InetSocketAddress(it, port) }

}
